### PR TITLE
chore: Deprecate old utility functions and minor type updates

### PR DESCRIPTION
**Pull request by NicholasBoll**
## Summary

Deprecate old utility functions and types:
- Types:
  - `Model`
  - `ToModelConfig`
- Functions:
  - `createEventMap`
  - `useEventMap`

All these utility types and functions have been replaced by `createModelHook`. Typescript 4.1 introduced [Template Literal Types](https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html), so event maps are no longer needed. `createModelHook` infers all types based on the input of the function.

Minor fixes to types

## Release Category
Utilities

---

## For the Reviewer

- [ ] PR title is short and descriptive
- [ ] PR summary describes the change (Fixes/Resolves linked correctly)
- [ ] PR Release Notes describes additional information useful to call out in a release message or removed if not applicable
- [ ] Breaking Changes provides useful information to upgrade to this code or removed if not applicable

### CODE REVIEW

File: modules/react/common/lib/utils/components.ts [hunk 1]
```diff
@@ -287,7 +296,7 @@ export const createContainer = <
           Context.Provider,
           {value: localModel},
           Component(
-            finalElemProps as Props & RemoveNull<ReturnType<TElemPropsHook>> & {ref: ExtractRef<E>},
+            finalElemProps as any,
```
Comment (mannycarrera4) [inline modules/react/common/lib/utils/components.ts:299]: why did this change to any?

File: modules/react/common/lib/utils/models.ts [hunk 2]
```diff
@@ -1,6 +1,9 @@
 /* eslint-disable react-hooks/rules-of-hooks */
 import React from 'react';
 
+/**
+ * @deprecated
```
Comment (mannycarrera4) [inline modules/react/common/lib/utils/models.ts:5]: should we add a message here as to what to use i guess not if they're using the new utilities

File: modules/react/common/lib/utils/components.ts [hunk 3]
```diff
@@ -303,9 +312,13 @@ export const createContainer = <
       // we'll cast to `Record<string, any>` for assignment. Note the lack of type checking
       // properties. Take care when changing the runtime of this function.
       (ReturnedComponent as Record<string, any>)[key] = (subComponents as Record<string, any>)[key];
+      // Add a displayName if one isn't already created. This prevents us from having to add
+      // `displayName` to subcomponents and only container components
```
Comment (NicholasBoll) [inline modules/react/common/lib/utils/components.ts:316]:
```suggestion
      // `displayName` to subcomponents if a container component has a `displayName`
```

File: modules/react/common/lib/utils/components.ts (hunk 1 above)
Comment (NicholasBoll) [inline modules/react/common/lib/utils/components.ts:299]:
```suggestion
            finalElemProps,
```

File: modules/react/common/lib/utils/components.ts [hunk 4]
```diff
@@ -348,17 +376,7 @@ export const createSubcomponent = <
...
-          ? TProps extends {ref: infer R}
-            ? TProps
-            : TProps & {ref: ExtractRef<E>}
-          : {ref: ExtractRef<E>}) &
-        (Props extends {children: any}
-          ? {}
-          : {
-              children?: React.ReactNode;
-            }),
+      props: CompoundProps<Props, TElemPropsHook, E>,
```
Comment (NicholasBoll) [inline modules/react/common/lib/utils/components.ts:379]: @mannycarrera4 I extracted out the prop interface into a type that has JSDoc for better understanding.

File: modules/react/common/lib/utils/components.ts [hunk 5]
```diff
@@ -303,9 +305,13 @@ export const createContainer = <
       // we'll cast to `Record<string, any>` for assignment. Note the lack of type checking
       // properties. Take care when changing the runtime of this function.
       (ReturnedComponent as Record<string, any>)[key] = (subComponents as Record<string, any>)[key];
+      // Add a displayName if one isn't already created. This prevents us from having to add
+      // `displayName` to subcomponents if a container component has a `displayName`
+      if (displayName && !(subComponents as Record<string, any>)[key].displayName) {
+        (subComponents as Record<string, any>)[key].displayName = `${displayName}.${key}`;
```
Comment (mannycarrera4) [inline modules/react/common/lib/utils/components.ts:311]: sneaky

File: modules/react/common/lib/utils/components.ts [hunk 6]
```diff
@@ -314,9 +320,30 @@ export const createContainer = <
...
 
-// export type ModelComponentFn<TModelHook, TElemPropsHook, SubComponents> = <Props={}>(Component: ())
+/**
+ * Props for the compound component based on props, elemPropsHook, and element. It will
+ * return a prop interface according to all these inputs. The following will be added to
+ * the passed in `Props` type:
+ * - The prop interface returned by the `elemPropsHook` function
+ * - if there is no detected `ref` in the `Props` interface, a `ref` will be added based on the element type E
+ * - if there is no detected `children` in the `Props` interface, `children` will be added based on `ReactNode`
+ */
```
Comment (mannycarrera4) [inline modules/react/common/lib/utils/components.ts:337]: thank you for explaining that

**Review by mannycarrera4 — DISMISSED** [review]

**Comment by mannycarrera4** [comment]
Add info to upgrade guide!

**Review by mannycarrera4 — APPROVED** [review]

**Review by mannycarrera4 — APPROVED** [review]

**Pull request merged**